### PR TITLE
LayerNorm

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes #{issue number}

### CODE REVIEW

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 1]
```diff
@@ -271,6 +272,76 @@ Operation createUnaryOp(
...
+  // enable_cudnn not used
+  pop(stack);
+  auto eps = pop(stack).toDouble();
+
+  // TODO: avoid doing these unwrappings
+  Tensor bias{};
+  Tensor weight{};
+  auto bias_ival = pop(stack);
+  if (!bias_ival.isNone()) {
+    bias = to_dense(bias_ival.toTensor());
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:300]: Why are we doing this ?

**Review by eellison — COMMENTED** [review]
Can you add some details to description here ? It's a bad sign to be doing layout conversions within the MKLDNN Subgraph

**Review by eellison — COMMENTED** [review]
If we're just converting all the inputs to dense and back to_mkldnn within the operator that is no better than not supporting the operator

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 2]
```diff
@@ -32,10 +32,57 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm_backward(
 
 #include <ATen/native/mkldnn/MKLDNNCommon.h>
 #include <ATen/native/mkldnn/Utils.h>
+#include <ATen/native/layer_norm.h>
 
 namespace at {
 namespace native {
 
+std::tuple<Tensor, Tensor, Tensor> mkldnn_layer_norm(
```
Comment (eellison) [inline aten/src/ATen/native/mkldnn/Normalization.cpp:40]: we're checking for certain preconditions in fusing layer_norm within jit like optional bias, optional weight, etc. i'm a little wary of adding ops here to aten that dont cover the full semantics. i dont know if we really want to maintain layer norm in aten generally.

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 3]
```diff
@@ -32,10 +32,57 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm_backward(
...
+    const Tensor& input,
+    IntArrayRef normalized_shape, const c10::optional<Tensor>& weight_opt /* optional */, const c10::optional<Tensor>& bias_opt /* optional */,
+    double eps) {
+
+  TORCH_CHECK(weight_opt.has_value(), "optional weight isn't supported");
+  TORCH_CHECK(bias_opt.has_value(), "optional bias isn't supported");
+  auto weight = *weight_opt;
+  auto bias = *bias_opt;
+  TORCH_CHECK(normalized_shape.size() == 1, "only accept shapes with one dimension");
```
Comment (eellison) [inline aten/src/ATen/native/mkldnn/Normalization.cpp:49]: another case where we are not covering full semantics...

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 4]
```diff
@@ -866,6 +908,14 @@ class MKLDNNSubgraphSlicer {
         return false;
       }
     }
+
+    if (n->matches(
+            "aten::layer_norm(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enable=True) -> Tensor") &&
+        n->namedInput("weight")->type() != NoneType::get() &&
+        n->namedInput("bias")->type() != NoneType::get()) {
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:915]: we also need a check on normalized_shape here

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 5]
```diff
@@ -32,10 +32,57 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm_backward(
...
+  TORCH_CHECK(normalized_shape[0] == input.sizes().back(), "the shape dimension should be equal to the last dimension of input");
```
Comment (eellison) [inline aten/src/ATen/native/mkldnn/Normalization.cpp:50]: wont this condition already be checked by `at::native::_check_layer_norm_inputs`? The quantization code does not include it https://github.com/pytorch/pytorch/blob/a9b0a921d592b328e7e80a436ef065dadda5f01b/aten/src/ATen/native/quantized/cpu/qnormalization.cpp#L26

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 6]
```diff
@@ -271,6 +273,32 @@ Operation createUnaryOp(
...
+
+  auto weight_ival = pop(stack);
+  TORCH_INTERNAL_ASSERT(weight_ival.isTensor());
+  weight = weight_ival.toTensor();
+
+  auto shape = pop(stack).toIntVector();
+  auto input = pop(stack).toTensor();
+
+  at::Tensor dst, mean, rstd;
+  std::tie(dst, mean, rstd) =
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:297]: so mean and variance here in the layer norm docs says: 
`Mean and variance are computed at runtime or provided by a user`. Could we provide this instead of computing on every run, is that possible? is it correct to initialize to zero ?

> For forward propagation, the mean and variance might be either computed at runtime (in which case they are outputs of the primitive) or provided by a user (in which case they are inputs). In the latter case, a user must set the dnnl_use_global_stats flag. 

are we sure we're doing this right ?

File: test/jit/test_freezing.py [hunk 7]
```diff
@@ -1927,6 +1927,23 @@ def test_pool3d_batchnorm(self):
...
+            N, C, H, W, = 10, 3, 224, 224
+            for param in ((model.conv1, W // 2, torch.randn(N, C, H, W)),
+                          (torch.nn.Linear(H, W), W, torch.randn(N, C, W)),):
+
+                for layernorm in (torch.nn.LayerNorm(param[1]), torch.nn.LayerNorm(param[1], elementwise_affine=False)):
+                    sub_model = torch.nn.Sequential(param[0], layernorm)
+                    sub_model.eval()
+                    mod = torch.jit.freeze(torch.jit.script(sub_model))
+                    self.run_pass("convert_frozen_ops_to_mkldnn", mod.graph)
+                    FileCheck().check_count("aten::to_dense", 1, exactly=True).run(mod.graph)
```
Comment (eellison) [inline test/jit/test_freezing.py:1944]: can we check that layer norrm is between to_mkldnn and to_dense

**Review by eellison — COMMENTED** [review]
Looking good, i had a few questions. Maybe we should rope in someone from intel to review as well.

**Review by eellison — COMMENTED** [review]
> Both forward and backward propagation support in-place operations, meaning that 
src can be used as input and output for forward propagation

this would be another reason to not have it as a built-in aten operator, so we can inplace it in ur fusion pass, since layer_norm_ doesnt exist in aten (might be a nice speedup)

**Comment by CaoZhongZ** [comment]
Generally good. I see y'all take care of the details.

The caveat of oneDNN LayerNorm is that it only support last dim Normalization. Substitution and runtime should be careful and defend about it. And there are in-place opportunities if input and output have the same data type width.

File: aten/src/ATen/native/mkldnn/Normalization.cpp (hunk 2 above)
Comment (Krovatkin) [reply aten/src/ATen/native/mkldnn/Normalization.cpp:40]: That's a good point. I will still need to add some layernorm helper to this file since I can't call into MKLDNN directly from `frozen_ops_to_mkldnn`

File: aten/src/ATen/native/mkldnn/Normalization.cpp (hunk 5 above)
Comment (Krovatkin) [reply aten/src/ATen/native/mkldnn/Normalization.cpp:50]: yeah, it does, I wanted a more specific error message though

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 6 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:297]: > Could we provide this instead of computing on every run, is that possible?

mean and variance are computed for the input tensor, though, not sure how we can avoid re-computation on every run unless they are computed somewhere else in a graph? 

> are we sure we're doing this right ? 

We aren't setting the flag and we are recomputing them, so we should be using those tensors solely as outputs.

File: aten/src/ATen/native/mkldnn/Normalization.cpp (hunk 2 above)
Comment (Krovatkin) [reply aten/src/ATen/native/mkldnn/Normalization.cpp:40]: Let me try if I can turn this into a helper

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 8]
```diff
@@ -32,10 +32,52 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm_backward(
...
 
 namespace at {
 namespace native {
 
+std::tuple<Tensor, Tensor, Tensor> mkldnn_layer_norm_last_index_weight_bias(
+    const Tensor& input,
+    IntArrayRef normalized_shape, const Tensor& weight, const Tensor& bias,
+    double eps) {
+
+  TORCH_INTERNAL_ASSERT(normalized_shape.size() == 1, "only accept shapes with the last dimension");
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Normalization.cpp:45]: MKLDNN only supports the normalization of the last index. But from the caller side, you can do itensor::reshape to coalesce the last dims into one before calling `ideep::layer_normalization_forward::compute`. Pasting some sample code FYI (not necessarily match the mkldnn tensor usage, just FYI):
```C++
std::tuple<at::Tensor, at::Tensor, at::Tensor> dil_native_layer_norm_impl(
    const at::Tensor &X, const at::Tensor &gamma /* optional */,
    const at::Tensor &beta /* optional */, int64_t M, int64_t N, double eps) {
  ideep::tensor x = itensor_view_from_dense(X);
  auto gamma_fp32 = gamma.to(at::kFloat);
  auto beta_fp32 = beta.to(at::kFloat);
  const ideep::tensor scale = itensor_view_from_dense(gamma_fp32);
  const ideep::tensor shift = itensor_view_from_dense(beta_fp32);
  int64_t i = 0;
  auto dim = at::maybe_wrap_dim(0, X.dim(), false);
  auto j = X.sizes()[dim];
  std::vector<int64_t> input_size;
  while (j <= M) {
    dim = at::maybe_wrap_dim(i++, X.dim(), false);
    input_size.push_back(X.sizes()[dim]);
    dim = at::maybe_wrap_dim(i, X.dim(), false);
    j *= X.sizes()[dim];
  }
  input_size.push_back(N);
  auto src = x.reshape(input_size);
  at::Tensor Y = at::native::empty_like(X);
  at::Tensor mean = at::empty({M}, X.options());
  at::Tensor variance = at::empty({M}, X.options());
  auto onednn_Y = itensor_view_from_dense(Y);
  auto onednn_mean = itensor_view_from_dense(mean);
  auto onednn_variance = itensor_view_from_dense(variance);
  ideep::layer_normalization_forward::compute(
      src, scale, shift, onednn_Y, onednn_mean, onednn_variance, eps);
  return std::make_tuple(Y, mean, variance);
}
```

**Comment by jgong5** [comment]
cc @liangan1 and @XiaobingSuper

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 9]
```diff
@@ -32,10 +32,52 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm_backward(
...
+  auto M_N = at::native::_check_layer_norm_inputs(input, normalized_shape, weight, bias);
+  auto M = M_N.first;
+
+  auto mean = empty_mkldnn(
+        {M},
+        input.scalar_type(),
```
[the remark this answers is not in the document]
Comment (eellison) [reply aten/src/ATen/native/mkldnn/Normalization.cpp:51]: It is an invariant within the fuser that all the inputs are float32 - @Krovatkin, I think we are making the scope of this PR much bigger by adding it in aten and it's not clear to me what the gain is, or at the very least I would just start with our more limited set of semantics and add to our fuser

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 6 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:298]: yea, we could try caching them for inference. that doesnt need to be done in the first version of this pr but it might be a nice freezing speed up. the mean and variance are wrt the weights right ?

Comment (Krovatkin) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:298]: unfortunately, it's wrt inputs :(
